### PR TITLE
Button and dialog for showing clients of a server in the main menu

### PR DESCRIPTION
**Pull request by siliconsniffer**
#### Goal of the PR
The goal of this PR is to add a button and a dialog in the main menu for showing the clients connected to a server.

#### How does the PR work?
It works by adding a button which when hovered over:
- shows the connected clients and appends an ellipsis if clients_list >= 5
- shows all connected clients if clients_list < 5
- shows a message "Clients not available" when the server doesn't "publish" the client names.

It also adds a dialog to the button showing:
- a label stating that "This is the list of clients from" + the name of the server
- a text list showing all connected clients from that server
- a button to exit the dialog

If the button is pressed or the dialog is exited by pressing Esc it also syncs the server list.

#### Does it resolve any reported issue?
Not any that I am aware of, this was/is a feature I wished we had for years at this point.

#### Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
I hope so.

#### If not a bug fix, why is this PR needed? What usecases does it solve?
Instead of having to connect to a server to e.g. check if a certain player is connected you can now retrieve this information from the main menu. This makes it easier when you want to meet with someone for example.

## To do
The plan is to have this as a foundation for a more general "info dialog" which could show all kinds of useful information. I also plan on fixing #15237 either by showing that in the future "info dialog" or, as proposed by sfan, by some sort of clickable label or similar. 

## How to test
Compile Luanti and open up the "Join Game" tab.
Click on a server and hover over the blue player icon.
Click on the button, a dialog opens up.
Watch how the server list sync when exiting from the dialog.

## Thanks
I want to thank:
- @appgurueu for the help during development
- @Zughy for the nice textures  and the great feedback.
- @jordan4ibanez for the motivational support :P
- everyone who gave feedback, really appreciate it!

## Preview
https://github.com/user-attachments/assets/e334f226-64e9-4a6d-8968-1684133d24c7

### CODE REVIEW

File: builtin/mainmenu/dlg_clients_list.lua [hunk 1]
```diff
@@ -0,0 +1,59 @@
+--Luanti
+--Copyright (C) 2024 siliconsniffer
+--
+--This program is free software; you can redistribute it and/or modify
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_clients_list.lua:4]: Suggestion: Use an SPDX-License-Identifier line to save some boilerplate.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+--This program is distributed in the hope that it will be useful,
+--but WITHOUT ANY WARRANTY; without even the implied warranty of
+--MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+--GNU Lesser General Public License for more details.
+--
+--You should have received a copy of the GNU Lesser General Public License along
+--with this program; if not, write to the Free Software Foundation, Inc.,
+--51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+--------------------------------------------------------------------------------
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_clients_list.lua:18]:
```suggestion
```

Please don't do these separators, they are a relict of 2014's coding style

File: builtin/mainmenu/dlg_clients_list.lua [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		"size[6,8]" ..
+		"hypertext[0.275,0;6,1;;<global halign=center>This is the list of clients from <b>" .. fgettext(servername) .. "]" ..
+		"textlist[0.4,0.5;5,7;;" .. fmt_formspec_list(clients_list) .. "]" ..
+		"button[1.5,7.6;3,0.8;dlg_clients_list_exit;OK]"
+end
+
+--------------------------------------------------------------------------------
+
+local function clients_list_buttonhandler(this, fields)
+	if fields.quit or fields["dlg_clients_list_exit"] then
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_clients_list.lua:42]:
```suggestion
	if fields.quit then
```

Just `fields.quit` suffices to detect formspec close. `button_exit` elements set it.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this:delete()
+        serverlistmgr.sync()
+		return true
+	end
+	return false
+end
+
+--------------------------------------------------------------------------------
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_clients_list.lua:50]:
```suggestion
```

File: builtin/mainmenu/tab_online.lua [hunk 5]
```diff
@@ -65,13 +65,15 @@ local function get_formspec(tabview, name, tabdata)
 	end
 
 	local retval =
-		-- Search
+	-- Search
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:68]: Ideally *try* to keep such unrelated changes ("noise") out of PRs. I don't mind a low level of noise / cleaning up adjacent code but it's better to do cleanups separately.

File: builtin/mainmenu/tab_online.lua [hunk 6]
```diff
@@ -65,13 +65,15 @@ local function get_formspec(tabview, name, tabdata)
...
 		"field[0.25,0.25;7,0.75;te_search;;" .. core.formspec_escape(tabdata.search_for) .. "]" ..
 		"field_enter_after_edit[te_search;true]" ..
 		"container[7.25,0.25]" ..
 		"image_button[0,0;0.75,0.75;" .. core.formspec_escape(defaulttexturedir .. "search.png") .. ";btn_mp_search;]" ..
-		"image_button[0.75,0;0.75,0.75;" .. core.formspec_escape(defaulttexturedir .. "clear.png") .. ";btn_mp_clear;]" ..
-		"image_button[1.5,0;0.75,0.75;" .. core.formspec_escape(defaulttexturedir .. "refresh.png") .. ";btn_mp_refresh;]" ..
+		"image_button[0.75,0;0.75,0.75;" ..
+		core.formspec_escape(defaulttexturedir .. "clear.png") .. ";btn_mp_clear;]" ..
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:74]: Such changes could for example very well be considered problematic: This makes it less obvious what belongs to this image button element. It would be good if it was at least indented.

File: builtin/mainmenu/tab_online.lua [hunk 7]
```diff
@@ -84,13 +86,13 @@ local function get_formspec(tabview, name, tabdata)
 		"label[0.25,0.35;" .. fgettext("Address") .. "]" ..
 		"label[4.25,0.35;" .. fgettext("Port") .. "]" ..
 		"field[0.25,0.5;4,0.75;te_address;;" ..
-			core.formspec_escape(core.settings:get("address")) .. "]" ..
+		core.formspec_escape(core.settings:get("address")) .. "]" ..
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:89]: Removing the indentation here is also a change for the worse IMO. Please check your editor configuration to make sure it isn't running some kind of aggressive auto formatter.

File: builtin/mainmenu/tab_online.lua [hunk 8]
```diff
@@ -111,13 +113,44 @@ local function get_formspec(tabview, name, tabdata)
...
-			retval = retval .. "image_button[5,1.3;0.5,0.5;" .. core.formspec_escape(defaulttexturedir ..
+			retval = retval .. "image_button[4.5,1.3;0.5,0.5;" .. core.formspec_escape(defaulttexturedir ..
 				"server_favorite_delete.png") .. ";btn_delete_favorite;]"
 		end
 		if gamedata.serverdescription then
 			retval = retval .. "textarea[0.25,1.85;5.25,2.7;;;" ..
 				core.formspec_escape(gamedata.serverdescription) .. "]"
 		end
+
+		local function shortenTable(clients_list)
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:124]: Should be `shorten_table` according to our naming conventions, see https://dev.minetest.net/Lua_code_style_guidelines

(also, this could be inlined at the only place where it is used. whether that is more readable or not may be debatable, but at least the `math.min` becomes obsolete then: You just have to pick the first 5 because you know that there are `>= 5`. actually, you can always pick the first 5 because `nil`s in Lua are funny but that wouldn't be very clean)

File: builtin/mainmenu/tab_online.lua [hunk 9]
```diff
@@ -111,13 +113,44 @@ local function get_formspec(tabview, name, tabdata)
...
+			return short_clients_list
+		end
+
+		local server = tabdata.lookup[tabdata.selected]
+
+		if server and server.clients_list then
+			table.sort(server.clients_list, function(a, b)
+				return string.lower(a) < string.lower(b)
+			end)
+			local clients_string = table.concat((server.clients_list), "\n")
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:138]:
```suggestion
			local clients_string = table.concat(server.clients_list, "\n")
```

File: builtin/mainmenu/tab_online.lua [hunk 10]
```diff
@@ -241,11 +274,11 @@ local function set_selected_server(tabdata, idx, server)
 		return
 	end
 
-	local address = server.address
-	local port    = server.port
+	local address              = server.address
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:277]: This level of alignment seems excessive, same below

File: builtin/mainmenu/tab_online.lua [hunk 11]
```diff
@@ -341,30 +382,29 @@ local function main_button_handler(tabview, fields, name, tabdata)
 	local te_port_number = tonumber(fields.te_port)
 
 	if (fields.btn_mp_login or fields.key_enter) and host_filled then
-		gamedata.playername = fields.te_name
-		gamedata.password   = fields.te_pwd
-		gamedata.address    = fields.te_address
-		gamedata.port       = te_port_number
+		gamedata.playername               = fields.te_name
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:385]: What's the matter with this change?

File: builtin/mainmenu/dlg_clients_list.lua [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		local escaped = {}
+		for i, str in ipairs(clients_list) do
+			escaped[i] = core.formspec_escape(str)
+		end
+		return table.concat(escaped, ",")
+	end
+
+	return
+		"size[6,8]" ..
+		"hypertext[0.275,0;6,1;;<global halign=center>This is the list of clients from <b>" .. fgettext(servername) .. "]" ..
```
Comment (appgurueu) [inline builtin/mainmenu/dlg_clients_list.lua:34]: This `<b>` tag is not closed?

`fgettext` looks up a translation in Luanti's translations, it doesn't make sense to apply it to a server name.

Something like `fgettext("This is the list of clients from $1", servername)` would make more sense. (I'm not sure whether `fgettext("This is the list of clients from $1", "<b>" .. servername .. "</b>")` works off the top of my head; try it and see.

Also, the server name should be hypertext-escaped. Otherwise we enable hypertext injection attacks. (For the most part it'll probably just be the ability to do funny styling or accidentally break the formspec but I'd prefer not to take any chances.)

File: builtin/mainmenu/tab_online.lua [hunk 13]
```diff
@@ -111,13 +113,44 @@ local function get_formspec(tabview, name, tabdata)
...
+			else
+				retval = retval .. "tooltip[btn_print_clients;" .. fgettext("Clients:\n$1", clients_string) .. "]"
+			end
+			retval = retval .. "style[btn_print_clients;padding=6]"
+			retval = retval .. "image_button[5,1.3;0.5,0.5;" .. core.formspec_escape(defaulttexturedir ..
+				"server_view_clients.png") .. ";btn_print_clients;]"
+		else
+			retval = retval .. "tooltip[btn_no_clients;" .. fgettext("Clients not available.") .. "]"
+			retval = retval .. "style[btn_no_clients;padding=6]"
+			retval = retval .. "image_button[5,1.3;0.5,0.5;" .. core.formspec_escape(defaulttexturedir ..
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:151]: I think this should probably be just an image. An image button wrongly suggests that you can click on it and something will happen.

**Review by appgurueu — COMMENTED** [review]
Code review only so far. Mostly style things. In terms of correctness, the hypertext / translation needs to be handled more carefully. Otherwise looks okay.

**Comment by grorp** [comment]
With the default window size (1024x600 iirc) and a long server name, it looks like this:

screenshot

The hypertext seems to continue underneath the textlist. Also, there seems to be excessive padding. Perhaps you could just remove all padding from the hypertext?

If you do vertical+horizontal vertical centering in the hypertext, removing padding shouldn't make a difference for large windows, while giving the text enough room in small windows.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	local function fmt_formspec_list(clients_list)
+		local escaped = {}
+		for i, str in ipairs(clients_list) do
+			escaped[i] = core.formspec_escape(str)
+		end
+		return table.concat(escaped, ",")
+	end
+
+	return
+		"size[6,8]" ..
```
Comment (grorp) [inline builtin/mainmenu/dlg_clients_list.lua:33]: Oh no, it seems you forgot to set a `formspec_version`, so you created the layout with the legacy formspec coordinate system! That's quite difficult and unpredictable compared to the current one...

Since the legacy formspec coordinate system has been legacy for a long time now (since 2019), adding a new formspec based on it doesn't make sense.

File: builtin/mainmenu/tab_online.lua [hunk 15]
```diff
@@ -210,13 +243,13 @@ local function search_server_list(input)
 		local name_matches, description_matches = true, true
 		for _, keyword in ipairs(keywords) do
 			name_matches = name_matches and not not
-					(server.name or ""):lower():find(keyword, 1, true)
+				(server.name or ""):lower():find(keyword, 1, true)
 			description_matches = description_matches and not not
-					(server.description or ""):lower():find(keyword, 1, true)
+				(server.description or ""):lower():find(keyword, 1, true)
```
[the remark this answers is not in the document]
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:248]: Yeah, only noticed it afterwards - must have been my editor. What's the best way to revert this?

File: builtin/mainmenu/tab_online.lua (hunk 13 above)
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:151]: I disagree. My intention was to keep the design with the "remove favorite" button consistent, in my opinion it would look off as just an image next to that button.

File: builtin/mainmenu/dlg_clients_list.lua (hunk 14 above)
Comment (siliconsniffer) [reply builtin/mainmenu/dlg_clients_list.lua:33]: That ... explains a lot. Fill add this, thanks for the hint.

File: builtin/mainmenu/tab_online.lua (hunk 15 above)
[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/mainmenu/tab_online.lua:248]: > What's the best way to revert this?

Unfortunately manual labor. First configure your editor properly.
Then look at the diff and fix the unrelated changes.
(Your editor should have some way to let you highlight the diff, and maybe directly let you undo changes for singular "hunks"; at least VSC lets me do that.)

**Comment by jordan4ibanez** [comment]
button

File: builtin/mainmenu/dlg_clients_list.lua (hunk 2 above)
Comment (siliconsniffer) [reply builtin/mainmenu/dlg_clients_list.lua]: Removed those.

**Comment by siliconsniffer** [comment]
Second image shows the dialog at 1024x600 with the longest server name I could find.
image
image

File: builtin/mainmenu/tab_online.lua (hunk 15 above)
[the remark this answers is not in the document]
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:244]: I tried reversing as  much as I can, what do you say?

File: builtin/mainmenu/dlg_clients_list.lua (hunk 1 above)
Comment (siliconsniffer) [reply builtin/mainmenu/dlg_clients_list.lua:4]: Would leave it like this for now for coherence with the other dialogs.

Comment (appgurueu) [reply builtin/mainmenu/dlg_clients_list.lua:4]: There is no need for consistency with legacy code in different files; in fact if anything I would consider it a (minor) oversight that we haven't gotten rid of this boilerplate in the legacy code yet.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	local function fmt_formspec_list(clients_list)
+		local escaped = {}
+		for i, str in ipairs(clients_list) do
+			escaped[i] = core.formspec_escape(str)
+		end
+		return table.concat(escaped, ",")
+	end
+
+	local formspec = {
+		"formspec_version[8]" ,
```
[the remark this answers is not in the document]
Comment (siliconsniffer) [reply builtin/mainmenu/dlg_clients_list.lua:32]: Ok, will do that tomorrow. Rest looks good?

File: builtin/mainmenu/dlg_clients_list.lua [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	return table.concat(formspec, "")
+end
+
+--------------------------------------------------------------------------------
+
+local function clients_list_buttonhandler(this, fields)
+	if fields.quit or fields["quit"] then
+        this:delete()
+        serverlistmgr.sync()
```
[the remark this answers is not in the document]
Comment (siliconsniffer) [reply builtin/mainmenu/dlg_clients_list.lua:48]: Yes, with "lower ranking" servers the order can and probably will change. I think it's fine though as the information of that server stays in the fields and it's easier than to require the user to refresh the server list themself. I don't see it as a necessity but rather a goodie :-)

[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/mainmenu/dlg_clients_list.lua:35]: I think refreshing the serverlist should still be explicit. I wouldn't expect viewing the clients of a server to refresh the serverlist.

[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/mainmenu/dlg_clients_list.lua]: https://github.com/minetest/minetest/pull/15425/commits/8acde060ba5dee4ef6ef4d9a621d796be1b95bfc

**Review by appgurueu — APPROVED** [review]
Should be good now. There still is an odd bug where the selected server index is off by one after exiting the dialog sometimes, but I don't see how it's related to the code at hand and think it is caused by something elsewhere. I'll investigate.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+		for i, str in ipairs(clients_list) do
+			escaped[i] = core.formspec_escape(str)
+		end
+		return table.concat(escaped, ",")
+	end
+
+	local formspec = {
+		"formspec_version[8]",
+		"size[6,9.5]",
+		"position[0.5,0.5]",
```
Comment (grorp) [inline builtin/mainmenu/dlg_clients_list.lua:21]: this is the default, no need to repeat it

File: builtin/mainmenu/dlg_clients_list.lua [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+		"hypertext[0,0;6,1.5;;<global margin=5 halign=center valign=middle>",
+			fgettext("This is the list of clients from \n$1",
```
Comment (grorp) [inline builtin/mainmenu/dlg_clients_list.lua:23]: very much a nitpick: trailing whitespace :)

```suggestion
			fgettext("This is the list of clients from\n$1",
```

File: builtin/mainmenu/dlg_clients_list.lua (hunk 18 above)
Comment (grorp) [reply builtin/mainmenu/dlg_clients_list.lua:21]: also, to make the formspec a bit nicer on mobile, I suggest adding

```lua
local TOUCH_GUI = core.settings:get_bool("touch_gui")
```

and

```suggestion
		TOUCH_GUI and "padding[0.01,0.01]" or "",
```

File: builtin/mainmenu/tab_online.lua [hunk 20]
```diff
@@ -311,7 +333,15 @@ local function main_button_handler(tabview, fields, name, tabdata)
 
 		serverlistmgr.delete_favorite(server)
 		-- the server at [idx+1] will be at idx once list is refreshed
-		set_selected_server(tabdata, idx, tabdata.lookup[idx+1])
+		set_selected_server(tabdata, idx, tabdata.lookup[idx + 1])
```
Comment (grorp) [inline builtin/mainmenu/tab_online.lua:336]: 1. unrelated formatting change (but might be acceptable since it's just one)
2. why isn't the comment the line above updated too, now it's inconsistent :(

Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua]: Formatted it according to the style of the comment.

File: builtin/mainmenu/dlg_clients_list.lua [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+		end
+		return table.concat(escaped, ",")
+	end
+
+	local formspec = {
+		"formspec_version[8]",
+		"size[6,9.5]",
+		TOUCH_GUI and "padding[0.01,0.01]" or "",
+		"hypertext[0,0;6,1.5;;<global margin=5 halign=center valign=middle>",
+			fgettext("This is the list of clients connected to\n$1",
```
Comment (sfan5) [inline builtin/mainmenu/dlg_clients_list.lua:24]: This string sounds awkward, should be something like "List of clients connected to $1:"

Comment (appgurueu) [reply builtin/mainmenu/dlg_clients_list.lua:24]: https://github.com/luanti-org/luanti/pull/15736#pullrequestreview-2604283778
